File: algorithm/code/Go/queue/arrayqueue.go
package queue

type ArrayQueue[T any] struct {
	data     []T
	capacity int // 队列容量
	front    int // 队头指针
	rear     int // 队尾指针
	size     int // 队列长度
}

type ArrayQueueMethod[T any] interface {
	Enqueue(T)
	Dequeue() (T, bool)
	Front() T
	IsEmpty() bool
	IsFull() bool
	Size() int
}

func NewArrayQueue[T any](capacity int) *ArrayQueue[T] {
	return &ArrayQueue[T]{
		data:     make([]T, capacity),
		capacity: capacity,
		front:    0,
		rear:     -1,
		size:     0,
	}
}

var _ ArrayQueueMethod[int] = (*ArrayQueue[int])(nil)

// 入队
func (aq *ArrayQueue[T]) Enqueue(data T) {
	// 判断队列是否已满
	if aq.IsFull() {
		return
	}
	aq.rear = (aq.rear + 1) % aq.capacity
	aq.data[aq.rear] = data
	aq.size++
}

// 出队
func (aq *ArrayQueue[T]) Dequeue() (T, bool) {
	// 判断队列是否为空
	if aq.IsEmpty() {
		var zero T
		return zero, false
	}
	// 队头元素出队
	item := aq.data[aq.front]
	// 更新对头
	aq.front = (aq.front + 1) % aq.capacity
	aq.size--
	return item, true
}

// 队头
func (aq *ArrayQueue[T]) Front() T {
	if aq.IsEmpty() {
		var zero T
		return zero
	}
	return aq.data[aq.front]
}

// 判断队列是否为空
func (aq *ArrayQueue[T]) IsEmpty() bool {
	return aq.size == 0
}

// 判断队列是否已满
func (aq *ArrayQueue[T]) IsFull() bool {
	return aq.size == aq.capacity
}

// 队列长度
func (aq *ArrayQueue[T]) Size() int {
	return aq.size
}

File: algorithm/code/TypeScript/queue/array-queue.ts
interface IArrayQueue<T> {
    enqueue(item: T): void
    dequeue(): T
    getFront(): T
    isEmpty(): boolean
    isFull(): boolean
    getSize(): number
}

export default class ArrayQueue<T> implements IArrayQueue<T> {
    private data: T[]
    private capacity: number
    private front: number
    private rear: number
    private size: number

    constructor(capacity: number = 10) {
        this.data = new Array(capacity)
        this.capacity = capacity
        this.front = 0
        this.rear = -1
        this.size = 0
    }

    enqueue(item: T) {
        if (this.isFull()) {
            throw new Error('queue is full')
        }
        this.rear = (this.rear + 1) % this.capacity
        this.data[this.rear] = item
        this.size++
    }

    dequeue(): T {
        if (this.isEmpty()) {
            throw new Error('queue is empty')
        }
        const data = this.data[this.front]
        this.front = (this.front + 1) % this.capacity
        this.size--
        return data
    }

    getFront() {
        if (this.isEmpty()) {
            throw new Error('queue is empty')
        }
        return this.data[this.front]
    }

    isFull() {
        return this.size === this.capacity
    }

    isEmpty() {
        return this.size === 0
    }

    getSize() {
        return this.size
    }
}